packages/eslint-plugin/src/rules/no-invalid-void-type.ts

```diff
@@ -174,6 +174,135 @@ export default createRule<[Options], MessageIds>({
       );
     }
 
+    function getMembers(
+      node: TSESTree.Node,
+    ):
+      | TSESTree.ClassElement[]
+      | TSESTree.ProgramStatement[]
+      | TSESTree.Statement[] {
+      switch (node.type) {
+        case AST_NODE_TYPES.ClassBody:
+        case AST_NODE_TYPES.Program:
+        case AST_NODE_TYPES.TSModuleBlock:
+        case AST_NODE_TYPES.BlockStatement:
+          return node.body;
+      }
+
+      /* istanbul ignore next */
+      return [];
```

Comment:
Is there a preferred way to tackle this? I think this is currently an unreachable line.

Reply:
Interesting! I didn't consider proposing changes/improvements to the AST types. I've found #9560, and I think it would be a good option if I could be sure the changes are correct.

To my understanding, TypeScript's AST doesn't even have an actual AST node for `ExportDefaultDeclaration` or `ExportNamedDeclaration` as they're expressed as [`modifiers`](https://github.com/microsoft/TypeScript/blob/f69580f82146bebfb2bee8c7b8666af0e04c7e34/src/compiler/types.ts#L2083).

I've looked into [`@babel/parser`'s types](https://github.com/babel/babel/blob/87a5e2eef28a1f5a3783da5ed25a163236541b8d/packages/babel-parser/src/types.ts), [`@swc/types`](https://github.com/swc-project/swc/blob/9eceec44dffdfef7eb4079bf35e82301383528c8/packages/types/index.ts), and [`@oxc-project/types`](https://github.com/oxc-project/oxc/blob/main/npm/oxc-types/types.d.ts), but I didn't find anything that makes me confident this is valid in every case. I did verify `@typescript-eslint` itself is type-checked successfully with these assumptions (also the more technical [1](https://tc39.es/ecma262/multipage/ecmascript-language-functions-and-classes.html#sec-function-definitions), [2](https://tc39.es/ecma262/multipage/ecmascript-language-scripts-and-modules.html#sec-exports)).

The stricter parent types for the AST I'd like to propose:
- `ExportDefaultDeclaration` and `ExportNamedDeclaration` should have the stricter `parent` type of `TSESTree.BlockStatement | TSESTree.Program | TSESTree. TSModuleBlock` (I'm not sure if `TSESTree.BlockStatement` is an option.
- `FunctionDeclaration` should have the stricter `parent` type of `TSESTree.BlockStatement | TSESTree.ExportDefaultDeclaration | TSESTree.ExportNamedDeclaration | TSESTree.Program`.
  
  TypeScript itself doesn't have a specific `parent` for `FunctionDeclaration` (it falls back to the generic `ts.Node`), although they do have stricter parent types for some other nodes.

How did we verify the correctness of the proposals in #6225? Also, would this type of change be considered a breaking change? I noticed that #9560 was released in v8, but it wasn't marked as a [breaking change](https://github.com/typescript-eslint/typescript-eslint/blob/main/CHANGELOG.md#-features-18).

I'm a bit hesitant to change the AST's types, as it can potentially lead to bugs, and I would love to hear opinions on this.

I updated the PR with these changes, so it's easier to discuss them.